### ISSUE TITLE
Circular dependencies between repos

### ISSUE THREAD

**Issue by reckart** [eclipse/rdf4j#1302]
It seems as if rdf4j depends on modules from rdf4j-storage (e.g. `rdf4j-repository-sail`) and of course rdf4j-storage depends on modules from rdf4j. So if I build them, they are always missing dependencies. Is there some trick?

**Comment by reckart** [comment]
I realize that I can disable the `compliance` profile on rdf4j using `-P!compliance` - but is that the way it is supposed to work?

**Comment by barthanssens** [comment]
Hi,

`mvn -Passembly clean install -DskipTests` should work, see also:

http://docs.rdf4j.org/developer/#_building_and_uploading_sdk_and_onejar

**Comment by reckart** [comment]
I have set up build jobs for rdf4j-storage and rdf4j on our Jenkins and the circular dependency causes endless builds there if the "build when snapshot dependency is updated" trigger is active. 

It would actually be nice if the full build would be possible without having to disable tests.

I have tried setting up a "super" POM which uses rdf4j, rdf4j-storage and rdf4j-tools as submodules, but even that didn't work because it appears as if there are modules by the same names in each of these repositories.

**Comment by jeenbroekstra** [comment]
@reckart as you've found out these are caused by test dependencies. It's a known issue and we're looking into it, hopefully soon once 2.5 is out the door. See also https://github.com/eclipse/rdf4j/pull/1264 which is intended as a temporary workaround (using fixed older versions of some of the dependencies instead of current project version). The real fix is to either stub/mock these test dependencies or else move the integration test (we're doing this on a case by case basis).

**Comment by jeenbroekstra** [comment]
Related issue: #1263

**Comment by jeenbroekstra** [comment]
Now that PR #1264 is merged, the immediate problem you're facing should be fixed: rdf4j still has (test) dependencies on some modules from rdf4j-storage, but it relies on a previously released version, rather than the current codebase version (e.g. 2.5-SNAPSHOT or 3.0-SNAPSHOT). This means that it should no longer be seen by Maven/Jenkins as a cyclic dependency. As said the better fix is to make sure we can remove these dependencies completely and instead just stub/mock what we need.

Let me know if the current setup works for you though.

EDIT I just noticed that you asked about disabling the compliance module. Yes, that is the way it is intended to work: build/install/deploy the project with compliance disabled, then run a second job that can use the installed artifacts to run verification with compliance tests. You can see how we've set this up (roughly) in our own Jenkins instance (at https://ci.eclipse.org/rdf4j): the `rdfj4-*-build` jobs all run a `mvn clean deploy -DskipTests` (which installs/deploys all builds to the snaphot repo, skipping compliance testing), a downstream followup `rdf4j-*-verify` job then executes full tests using `mvn integration-test -Pcompliance`.

**Comment by jeenbroekstra** [comment]
Out of curiosity: why are you setting up your own Jenkins build for rdf4j?

**Comment by jeenbroekstra** [comment]
also related: #1236 . We're looking to simplify the setup of compliance and unit tests more generally.

**Comment by reckart** [comment]
> Out of curiosity: why are you setting up your own Jenkins build for rdf4j?

Because for our [project](https://inception-project.github.io), we needed to merge the Lucene patch (https://github.com/eclipse/rdf4j-storage/pull/78 - actually compiling against Lucene 7.3.0, we don't need the modules which are broken by the patch) and we also may need other custom changes such as (https://github.com/eclipse/rdf4j/pull/1303) where we are not sure how quickly they can be available in an official RDF4J release. We also might look into an attempt to extract human-readable names from IRIs and allowing the FTS to index these (https://github.com/inception-project/rdf4j-storage/issues/3).

**Comment by jeenbroekstra** [comment]
@reckart I see, looks like an interesting project! Fwiw we have pretty quick turnaround on patch releases, so for bug fixes at least, if you're going through the effort of creating a fix anyway, you might as well push them back to us. Even if we don't immediately do a release, we build daily snapshots that are deployed to Sonatype, so there'll be an artifact with the fix available as soon as a PR is merged.

The main reason the lucene upgrade is held up is pretty much that no-one currently has time to run full tests - and we didn't want it to block the 2.5 release. Given that you have an interest in it, we'd love your feedback on the current state of it.

**Comment by reckart** [comment]
If I remember correctly, when we merged the patch, we had to comment out some modules which didn't build anymore (e.g. some related to elastic search). I don't know if there was any update to 
https://github.com/eclipse/rdf4j-storage/pull/78 which might have fixed that. Otherwise, it seems to work pretty much as expected ;)

We had to merge the Lucene PR because we have another library (MTAS) depending on Lucene as well and needed to resolve an incompatible version conflict on Lucene.
